lightning-c-bindings/src/c_types/mod.rs

```diff
@@ -0,0 +1,415 @@
+pub mod derived;
+
+use bitcoin::Script as BitcoinScript;
+use bitcoin::Transaction as BitcoinTransaction;
+use bitcoin::secp256k1::key::PublicKey as SecpPublicKey;
+use bitcoin::secp256k1::key::SecretKey as SecpSecretKey;
+use bitcoin::secp256k1::Signature as SecpSignature;
+
+#[derive(Clone)]
+#[repr(C)]
+pub struct PublicKey {
+	pub compressed_form: [u8; 33],
+}
+impl PublicKey {
+	pub(crate) fn from_rust(pk: &SecpPublicKey) -> Self {
+		Self {
+			compressed_form: pk.serialize(),
+		}
+	}
+	pub(crate) fn into_rust(&self) -> SecpPublicKey {
+		SecpPublicKey::from_slice(&self.compressed_form).unwrap()
+	}
+	pub(crate) fn is_null(&self) -> bool { self.compressed_form[..] == [0; 33][..] }
+	pub(crate) fn empty() -> Self { Self { compressed_form: [0; 33] } }
+}
+
+#[repr(C)]
+pub struct SecretKey {
+	pub bytes: [u8; 32],
+}
+impl SecretKey {
+	// from_rust isn't implemented for a ref since we just return byte array refs directly
+	pub(crate) fn from_rust(sk: SecpSecretKey) -> Self {
+		let mut bytes = [0; 32];
+		bytes.copy_from_slice(&sk[..]);
+		Self { bytes }
+	}
+	pub(crate) fn into_rust(&self) -> SecpSecretKey {
+		SecpSecretKey::from_slice(&self.bytes).unwrap()
+	}
+}
+
+#[repr(C)]
+pub struct Signature {
+	pub compact_form: [u8; 64],
+}
+impl Signature {
+	pub(crate) fn from_rust(pk: &SecpSignature) -> Self {
+		Self {
+			compact_form: pk.serialize_compact(),
+		}
+	}
+	pub(crate) fn into_rust(&self) -> SecpSignature {
+		SecpSignature::from_compact(&self.compact_form).unwrap()
+	}
+	pub(crate) fn is_null(&self) -> bool { self.compact_form[..] == [0; 64][..] }
+	pub(crate) fn empty() -> Self { Self { compact_form: [0; 64] } }
+}
+
+#[repr(C)]
+/// A reference to a serialized transaction, in (pointer, length) form.
+/// This type does *not* own its own memory, so access to it after, eg, the call in which it was
+/// provided to you are invalid.
+pub struct Transaction {
+	pub data: *const u8,
+	pub datalen: usize,
+}
+impl Transaction {
+	pub(crate) fn into_bitcoin(&self) -> BitcoinTransaction {
+		// datalen == 0 is invalid, so don't bother checking here.
```

Comment:
What does it mean invalid? the `from_slice` method can def generate that.
the next line is fine if `datalen = 0` as long as `data` is an aligned pointer, but if it is null it is not fine.

See: https://doc.rust-lang.org/beta/std/slice/fn.from_raw_parts.html#safety
> `data` must be valid for reads for len * mem::size_of::<T>() many bytes, and it must be properly aligned

> data must be non-null and aligned even for zero-length slices. One reason for this is that enum layout optimizations may rely on references (including slices of any length) being aligned and non-null to distinguish them from other data. You can obtain a pointer that is usable as data for zero-length slices using NonNull::dangling()..

Reply:
Its invalid in the sense that an empty slice/buffer can never represent a valid bitcoin transaction. Still, I added an explicit panic since that is a more useful error message.